library/eddsa.c

```diff
@@ -0,0 +1,434 @@
+
+cleanup:
+    mbedtls_mpi_free( &q );
+    mbedtls_mpi_free( &prefix );
+    mbedtls_mpi_free( &rq );
+    mbedtls_mpi_free( &h );
+    mbedtls_ecp_point_free( &Q );
+    mbedtls_ecp_point_free( &R );
+
+    return( ret );
+
+}
+
+int mbedtls_eddsa_verify( mbedtls_ecp_group *grp,
+                          const unsigned char *buf, size_t blen,
+                          const mbedtls_ecp_point *Q, const mbedtls_mpi *r,
+                          const mbedtls_mpi *s,
+                          mbedtls_eddsa_id eddsa_id,
+                          const unsigned char *ed_ctx, size_t ed_ctx_len,
+                          int (*f_rng)(void *, unsigned char *, size_t), void *p_rng )
+{
+    int ret = 0;
+    mbedtls_mpi h;
+    mbedtls_ecp_point sB, hA;
+
+    mbedtls_mpi_init( &h );
+    mbedtls_ecp_point_init( &sB );
+    mbedtls_ecp_point_init( &hA );
+
+    /* Step 1 */
+    if( mbedtls_mpi_cmp_mpi( s, &grp->N ) >= 0 || mbedtls_mpi_cmp_int( s, 0 ) < 0 )
+        return( MBEDTLS_ERR_ECP_BAD_INPUT_DATA );
+
+#ifdef MBEDTLS_ECP_DP_ED25519_ENABLED
+    if( grp->id == MBEDTLS_ECP_DP_ED25519 )
+    {
+        mbedtls_sha512_context sha_ctx;
+        unsigned char sha_buf[64], tmp_buf[32];
+        size_t olen = 0;
+
+        mbedtls_sha512_init( &sha_ctx );
+        mbedtls_sha512_starts( &sha_ctx , 0 );
+
+        /* Step 2 */
+        if( eddsa_id == MBEDTLS_EDDSA_CTX )
+        {
+            MBEDTLS_MPI_CHK( mbedtls_eddsa_put_dom2_ctx( 0, ed_ctx, ed_ctx_len, &sha_ctx ) );
+        }
+        else if( eddsa_id == MBEDTLS_EDDSA_PREHASH )
+        {
+            MBEDTLS_MPI_CHK( mbedtls_eddsa_put_dom2_ctx( 1, ed_ctx, ed_ctx_len, &sha_ctx ) );
+        }
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_write_binary_le( r, tmp_buf, sizeof(tmp_buf) ) );
+        mbedtls_sha512_update( &sha_ctx, tmp_buf, sizeof( tmp_buf ) );
+
+        MBEDTLS_MPI_CHK( mbedtls_ecp_point_write_binary( grp, Q, MBEDTLS_ECP_PF_COMPRESSED, &olen, tmp_buf, sizeof( tmp_buf ) ) );
+        mbedtls_sha512_update( &sha_ctx, tmp_buf, sizeof( tmp_buf ) );
+        mbedtls_platform_zeroize( tmp_buf, sizeof( tmp_buf ) );
+
+        /* In EDDSA_PREHASH, buf should contain the SHA512 hash. It contains the whole message otherwise */
+        mbedtls_sha512_update( &sha_ctx, buf, blen );
+
+        mbedtls_sha512_finish( &sha_ctx, sha_buf );
+        mbedtls_sha512_free( &sha_ctx );
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_read_binary_le( &h, sha_buf, sizeof( sha_buf ) ) );
+        mbedtls_platform_zeroize( sha_buf, sizeof( sha_buf ) );
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_mod_mpi( &h, &h, &grp->N ) );
+
+        /* Step 3 */
+        /* We perform fast single-signature verification by compressing sB-hA and comparing with r without decompressing it (expensive) */
+        MBEDTLS_MPI_CHK( mbedtls_ecp_mul( grp, &sB, s, &grp->G, f_rng, p_rng ) );
+        MBEDTLS_MPI_CHK( mbedtls_ecp_mul( grp, &hA, &h, Q, f_rng, p_rng ) );
+        MBEDTLS_MPI_CHK( mbedtls_ecp_sub( grp, &sB, &sB, &hA ) );
```

Comment:
You don't need to add code to subtract a point. Just negating `h` before the multiplication gives you the correct result.

Reply:
I did
```
        h.s *= -1;
        MBEDTLS_MPI_CHK( mbedtls_ecp_mul( grp, &hA, &h, Q, f_rng, p_rng ) );
        MBEDTLS_MPI_CHK( mbedtls_ecp_add( grp, &sB, &sB, &hA ) );
```
but it fails when `mbedtls_mpi_cmp()`. Am I missing something else?

Reply:
This is modular arithmetic. So just subtracts it from N:

`MBEDTLS_MPI_CHK( mbedtls_mpi_sub_mpi( &h, &grp->N, &h ) );`